gluonnlp/model/language_model.py

```diff
@@ -421,3 +423,95 @@ def standard_lstm_lm_1500(dataset_name=None, vocab=None, pretrained=False, ctx=c
+        Number of sampled classes for the decoder.
+    dropout : float
+        Dropout rate to use for encoder output.
+
+    """
+    def __init__(self, vocab_size, embed_size, hidden_size, num_layers,
+                 projection_size, dropout=0.0, **kwargs):
+        super(BigRNN, self).__init__(**kwargs)
+        self._embed_size = embed_size
+        self._hidden_size = hidden_size
+        self._projection_size = projection_size
+        self._num_layers = num_layers
+        self._dropout = dropout
+        self._vocab_size = vocab_size
+
+        with self.name_scope():
+            self.embedding = self._get_embedding()
+            self.encoder = self._get_encoder()
+            self.decoder = self._get_decoder()
+
+    def _get_embedding(self):
+        prefix = 'embedding0_'
+        embedding = nn.Sequential()
+        with embedding.name_scope():
+            embedding.add(nn.Embedding(self._vocab_size, self._embed_size, prefix=prefix))
+            if self._dropout:
+                embedding.add(nn.Dropout(self._dropout))
+        return embedding
+
+    def _get_encoder(self):
+        block = rnn.SequentialRNNCell()
+        with block.name_scope():
+            for _ in range(self._num_layers):
+                block.add(contrib.rnn.LSTMPCell(self._hidden_size, self._projection_size))
+                if self._dropout:
+                    block.add(rnn.DropoutCell(self._dropout))
+        return block
+
+    def _get_decoder(self):
+        output = nn.Dense(self._vocab_size, prefix='decoder0_')
+        return output
+
+    def begin_state(self, **kwargs):
+        return self.encoder.begin_state(**kwargs)
+
+    def forward(self, inputs, begin_state): # pylint: disable=arguments-differ
+        """Implement forward computation.
+
+        Parameters
+        -----------
+        inputs : NDArray
+            input tensor with shape `(sequence_length, batch_size)`
+            when `layout` is "TNC".
+        begin_state : list
+            initial recurrent state tensor with length equals to num_layers*2.
+            For each layer the two initial states have shape `(batch_size, num_hidden)`
+            and `(batch_size, num_projection)`
+
+        Returns
+        --------
+        out : NDArray
+            output tensor with shape `(sequence_length*batch_size, vocab_size)`
+              when `layout` is "TNC".
+        out_states : list
+            output recurrent state tensor with length equals to num_layers*2.
+            For each layer the two initial states have shape `(batch_size, num_hidden)`
+            and `(batch_size, num_projection)`
+        """
+        encoded = self.embedding(inputs)
+        length = inputs.shape[0]
+        encoded, state = self.encoder.unroll(length, encoded, begin_state,
+                                             layout='TNC', merge_outputs=True)
+        encoded = encoded.reshape((-1, self._projection_size))
+        out = self.decoder(encoded)
+        return out, state
```

Comment:
Reshape back to TNC? (Doesn't cost computation, but better matches user expectations)